File: baekjoon/solve/11049.c
/* Solution code for "BOJ 11049. Çà·Ä °ö¼À ¼ø¼­".

- Problem link: https://www.acmicpc.net/problem/11049
- MY Solution link: https://www.acmicpc.net/source/43454670
- Solution link:
*/
#include <stdio.h>
#define INF 1000000000
#define MAXSIZE 501
int min(int a, int b)
{
    return a < b ? a : b;
}

int main()
{
    int N;
    int start, end, v;
    int d[MAXSIZE];
    int dp[MAXSIZE][MAXSIZE] = {0};

    scanf("%d", &N);
    scanf("%d %d", &d[0], &d[1]);
    for (int i = 0; i < N - 1; i++)
    {
        scanf("%d %d", &d[i + 1], &d[i + 2]);
    }
    for (int distance = 2; distance < N + 1; distance++)
    {
        for (int i = 0; i < N - distance + 1; i++)
        {
            start = i;
            end = i + distance;
            dp[start][end] = INF;
            for (int m = start + 1; m < end; m++)
            {
                v = dp[start][m] + dp[m][end] + d[start] * d[m] * d[end];
                if (v < dp[start][end])
                {
                    dp[start][end] = v;
                }
            }
        }
    }
    printf("%d\n", dp[0][N]);
}

File: BOJ/solve/11049.py
"""TIME OVER code for "BOJ 11049. Çà·Ä °ö¼À ¼ø¼­".

- Problem link: https://www.acmicpc.net/problem/11049
- MY link: https://www.acmicpc.net/source/43454487
- Solution link: http://www.teferi.net/ps/problems/boj/11049
- Solution: (+ zip) -> time reduce
"""

INF = 1e12

N = int(input())
d = list(map(int, input().split()))
dp = [[0] * (N+1) for _ in range(N+1)]

for _ in range(N-1):
    d.append(list(map(int, input().split()))[1])

for distance in range(2, N+1):
    for i in range(N-distance+1):
        start = i
        end = i + distance

        dp[start][end] = INF

        for m in range(start+1, end):
            v = dp[start][m] + dp[m][end] + d[start] * d[m] * d[end]
            if v < dp[start][end]:
                dp[start][end] = v

print(dp[0][-1])
